Feat/step3

toongri opened the pull request.

toongri committed:
gitignore 최상단 out 디렉토리 추가

toongri committed:
subway 비활성화 컨트롤러 함수명 변경

toongri committed:
line 영속성 모델 시작역 id 업데이트로직 추가

toongri committed:
'삭제' 용어 -> '비활성화'로 수정

toongri committed:
out파일 삭제

toongri committed:
sectionAdd, sectionClose 싱글톤 패턴 -> 객체 생성으로 변경

toongri committed:
경로 탐색 추상화

PathSearcher 정의
PathSearcher/search 함수 정의

toongri committed:
지하철 구간 모델 하차역 비교 기능 정의

SubwaySection/isSameDownStation 함수 정의

toongri committed:
경로 탐색 역 모델 정의

PathStation 정의

toongri committed:
경로 탐색 구간 모델 정의

PathSection 정의

toongri committed:
경로 탐색 구간 JGraph 의존 모델 정의

PathSectionJGraph 정의

toongri committed:
거리 단위 모델 double 생성 기능 정의

Kilometer/of 정적 함수 정의

toongri committed:
지하철 구간 추가 요청 검증 테스트 수정

SubwaySectionAddCommandValidatorTest 수정

toongri committed:
경로 탐색 역 컬렉션 모델 정의

PathStations 정의

toongri committed:
경로 모델 정의

PathSubway 정의

toongri committed:
경로 탐색 최단거리 기능 정의

PathShortestDistanceSearcher 정의

toongri committed:
경로 탐색 최단거리 기능 테스트

PathShortestDistanceSearcherTest 정의

toongri committed:
지하철 역 영속성 모델 정적 생성 함수 정의

StationJpaMapper from 함수 수정
StationJpa 생성자 접근제어자 수정
StationJpa 정적 생성 함수 정의

toongri committed:
지하철 구간 영속성 모델 정적 생성 함수 생성

SubwaySectionJpa/constructor 접근제어자 수정
SubwaySectionJpa/register 정적 생성 함수 생성
SubwayLineJpa/addSections 수정
SubwayLinJpaMapper/from 수정
SubwayLineResponseMapperTest 수정

toongri committed:
지하철 구간 검증 로직 수정

SubwaySectionAddCommandValidator/validate 수정
SubwaySectionCloseCommandValidator/validate 수정
SubwayLine.Id/isEmpty 추가
Station.Id/isEmpty 추가

toongri committed:
경로 내 역 식별자 도메인 모델 null 체크 로직 추가

PathStation.Id/isEmpty 추가

toongri committed:
경로 내 지하철역 도메인 모델 조회 기능 추상화

PathStationsLoadPort 추가
PathStationsLoadPort/findAllBy 추가
PathStationsLoadPort.Response 추가

toongri committed:
경로 내 지하철역 도메인 모델 변환 기능 구현

PathStationMapper 추가

toongri committed:
경로 내 지하철역 도메인 모델 변환 기능 테스트

PathStationMapperTest 추가

toongri committed:
경로 내 지하철역 도메인 모델 조회 기능 구현

PathStationsLoadPersistenceAdapter 추가
PathStationsLoadPersistenceAdapter/findAllBy 추가

toongri committed:
경로 내 지하철역 도메인 모델 조회 기능 테스트

PathStationsLoadPersistenceAdapterTest 추가

toongri committed:
경로 검색 도메인 모델 조회 기능 추상화

PathSearcherLoadPort 추가

toongri committed:
지하철구간 영속성 모델 조회 기능 추상화

SubwaySectionJpaRepository 추가

toongri committed:
경로 구간 도메인 모델 변환 기능 구현

PathSectionMapper 추가
PathSectionMapper/from 추가

toongri committed:
경로 구간 도메인 모델 변환 기능 테스트

PathSectionMapperTest 추가

toongri committed:
경로 검색 도메인 모델 조회 기능 구현

PathSearcherLoadPersistenceAdapter 추가
PathSearcherLoadPersistenceAdapter/findOne 추가

toongri committed:
경로 검색 도메인 모델 조회 기능 테스트

PathSearcherLoadPersistenceAdapterTest 추가

toongri committed:
경로검색 명령 검증 기능 구현

PathQueryCommandValidator 추가
PathQueryCommandValidator/validate 추가

toongri committed:
경로 뷰 모델 구현

PathResponse 추가
PathResponse.Station 추가

toongri committed:
경로 검색 유즈케이스 추상화

PathQuery 추가
PathQuery/findOne 추가
PathQuery.Command 추가

toongri committed:
경로 뷰 모델 변환 기능 구현

PathResponseMapper 추가
PathResponseMapper/from 추가

toongri committed:
경로 검색 유즈케이스 구현

PathQueryService 추가
PathQueryService/findOne 추가

toongri committed:
경로 검색 유즈케이스 테스트

PathQueryServiceTest 추가

toongri committed:
경로 검색 컨트롤러 구현

PathQueryController 추가

toongri committed:
경로 검색 api 테스트 기능 구현

PathSearchApiTester 추가

toongri committed:
경로 검색 인수 테스트

PathSearchAcceptanceTest 추가

bactoria commented on src/main/java/subway/domain/SectionAddManager.java:
```diff
@@ -1,18 +1,14 @@
-import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Component;
 
 @Component
```
도메인 패키지에서 스프링 프레임워크에 대한 의존성이 들어가 있네요.

해당 클래스는 스프링 프레임워크에 대한 의존성이 불가피하다고 판단하신걸까요? 🤔

toongri replied:
불가피하다고 판단하지는 않았습니다.

합리화를 좀 했던 것 같습니다.

빠르게 진행하기 위해 이정돈 커버가 된다는 식의...

빠르게 수정해보겠습니다

bactoria commented on src/test/java/subway/domain/PathShortestDistanceSearcherTest.java:
```diff
@@ -0,0 +1,327 @@
+     * @when A역과 D역을 기준으로 경로탐색을 할 때
+     * @then 서로 다른 두 노선을 환승하여 최단 거리로 갈 수 있다.
+     */
+    @Test
+    @DisplayName("서로 다른 두 노선을 환승하여 최단 거리로 갈 수 있다.")
+    void searchPathBetweenTwoLines() {
+        //given
+        PathStation.Id 서울대입구역Id = PathStation.Id.of(1L);
+        PathStation 서울대입구역 = PathStation.of(서울대입구역Id, "서울대입구역");
+
+        PathStation.Id 강남역Id = PathStation.Id.of(2L);
+        PathStation 강남역 = PathStation.of(강남역Id, "강남역");
+
+        PathStation.Id 성수역Id = PathStation.Id.of(3L);
+        PathStation 성수역 = PathStation.of(성수역Id, "성수역");
+
+        PathStation.Id 건대입구역Id = PathStation.Id.of(4L);
+        PathStation 건대입구역 = PathStation.of(건대입구역Id, "건대입구역");
+
+        //given
+        PathSection.Id firstSectionId = PathSection.Id.of(1L);
+        PathSection firstSection = PathSection.of(
+                firstSectionId,
+                서울대입구역,
+                강남역,
+                Kilometer.of(5));
```
테스트 코드도 가독성이 중요하다고 생각하는데요!
아래 두가지 방법을 고려하여 given 부분의 라인수를 줄여보면 어떨까요?

- beforeEach 활용
- 메서드 추출 활용

역에 대한 생성 로직은 해당 테스트에 중요한 부분이 아니라 드러나지 않아도 상관없다고 생각해요.
이는 `@beforeEach` 를 활용해볼 수 있을 것 같아요.

반면, 구간의 거리는 중요한 정보니 메서드 추출을 이용한다면 테스트 코드의 가독성과 유지보수성이 올라갈 것 같아요.
```java
PathSection firstSection = createPathSection(1L, 서울대입구역, 강남역, Kilometer.of(5));
```

toongri replied:
아 제가 테스트 조회 시에 반복되는 코드를 줄이는 것보다 구현부가 최대한 서술되어야 한다는 생각에 조금 과했던 것 같습니다

말씀해주신 부분 읽고 경로 내 역 도메인 모델 생성 부분이 무분별하게 많다고 느꼈고, setUp함수로 뺐습니다.
또한 구간 도메인 모델은 말씀하신 것처럼 beforeEach로 빼면 안된고 구현부를 자세히 서술해줘야한다고 생각했지만 간단한 생성로직은 함수로 빼도 스크롤을 많이 쓰지 않아도 될 것 같아 말씀하신대로 수정해보았습니다

감사합니다!

bactoria commented on src/test/java/subway/domain/PathShortestDistanceSearcherTest.java:
```diff
@@ -0,0 +1,327 @@
+@DisplayName("노선 최단 거리 탐색에 대한 도메인 단위 테스트")
+class PathShortestDistanceSearcherTest {
+
+    /**
+     * @given A, B, C, D 역이 존재하고
+     * @given 탑승역이 A역, 하차역이 B역, 거리가 5인 구간이 존재하고
+     * @given 탑승역이 B역, 하차역이 C역, 거리가 12인 구간이 존재하고
+     * @given 탑승역이 A역, 하차역이 D역, 거리가 5인 구간이 존재하고
+     * @given 탑승역이 D역, 하차역이 C역, 거리가 8인 구간이 존재한다고 하면
+     * @when 두가지 경로를 갖고 있는 A역과 C역을 기준으로 경로탐색을 할 때
+     * @then 두 경로 중 총 거리가 더 적은 경로로 갈 수 있다.
+     */
```
상세한 인수조건 💯

bactoria commented on src/test/java/subway/acceptance/PathSearchAcceptanceTest.java:
```diff
@@ -0,0 +1,111 @@
+        신림역 = 지하철_역_생성("신림역").as(StationResponse.class);
+    }
+
+    /**
+     * @given A-B 구간을 갖고 있는 AB 노선이 존재하고
+     * @given B-C-D 구간을 갖고 있는 BCD 노선이 존재한다고 한다면
+     * @when 같은 B역을 기준으로 경로탐색을 할 때
+     * @then 에러 메시지를 출력한다.
+     */
```
예외케이스도 잘 테스트 해 주셨네요!
존재하지 않는 역에 대해서 경로 조회를 하는 경우에도 인수테스트를 해보면 어떨까요?

toongri replied:
제가 이번에 인수테스트를 여러번 반복하며 실행해본 결과 인수 테스트가 너무 디테일하게 될 경우에 속도면에서 큰 단점이 있다라는 것을 조금 인지했습니다.

맨 처음엔 정말 세세하게 모든 것에 대해 서술하였는데 가감을 해야하다보니 어떤 것을 살리고 어떤 것을 제거해야할까? 고민을 해봤습니다

고민의 결과는 해당 로직을 생각했을 때 어 이것은 어떻게 될까? 하고 사람들마다 예측이 분분할만한 것에 대해 인수 테스트를 작성해보자로 결론을 지었습니다.

그런 의미에서 존재하지 않는 역에 대한 경로 조회 경우는 상식적으로 대부분 아 에러가 발생되지 않을까? 라고 예측할 것이라고 저는 생각해서 인수 테스트를 작성해보지 않았습니다.

리뷰어님께서 생각해보시기에는 존재하지 않는 역에 대해 경로 조회를 하는 경우를 인수 테스트를 했을 때의 장점은 어떤것이라고 생각하는지 제가 여쭤봐도 될까요?

bactoria replied:
> 상식적으로 대부분 아 에러가 발생되지 않을까? 라고 예측할 것

경로 조회 시 역이 존재하지 않는 경우에는 따로 처리를 하지 않는 경우 500 에러가 발생할 것 같네요.

개발하는 입장에서는 어느정도 확신이 들겠지만, 
개발을 잘 모르는 기획자/테스터 의 시각에서도 확신을 가지고 있을까요? 🤔

<br>

마지막에 질문주신 부분은 인수조건을 모두 테스트로 구현하느냐 아니냐에 대한 질문으로 이해했는데요.

이번 미션에서 힌트로 가이드가 되어있어서 의견을 남겨두었어요.

<img width="339" alt="CleanShot 2023-09-08 at 07 37 55@2x" src="https://github.com/next-step/atdd-subway-path/assets/25674959/0089d297-9559-46ba-893f-da5efb8e2550">

하지만 실무에서는 모든 인수조건에 대해 인수테스트를 만들면 공수가 많이들어 
해당 케이스는 테스트하지 않는 것으로 진행 할 수도 있을 것 같습니다.

toongri replied:
확인했습니다. 제가 지금 밖이라 집와서 바로 추가하도록 하겠습니다!

bactoria commented on src/main/java/subway/domain/PathSubway.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * 경로 탐색 결과
+ */
+public class PathSubway {
```
지하철 이라는 객체가 있었나! 하는 생각에 흠칫 헀어요

SubwayPath 로 변경해보면 어떨까요? 😀

toongri replied:
통일성을 주려고 해보았는데 실패했네요 수정하겠습니다

bactoria commented on src/main/java/subway/application/mapper/PathResponseMapper.java:
```diff
@@ -0,0 +1,24 @@
+        return PathResponse.of(from(path.getStations()), path.getDistance());
+    }
+
+    private List<PathResponse.Station> from(List<PathStation> stations) {
```
Port & Adapter 에서 기본적으로 Mapper는 
Adapter 객체(eg. ORM 엔티티) 와 도메인 객체 간의 매핑을 담당하는 것으로 알고 있습니다.

domain 객체에서 다른 domain 객체로의 전환도 Mapper 를 만들어야 할까요? 🧐

toongri replied:
아 Response라는 이름이 담긴 친구들은 비즈니스 레이어가 담기지 않은 응답용 모델들입니다.

이 친구들도 도메인 모델이라고 봐야할까요?
조회용 모델들을 저는 서비스 레이어에 있으면 된다 생각했는데 리뷰어님이 생각하시기에 어떤지 궁금합니다

그리고 Mapper클래스는 저는 persistance 모델과 도메인 모델의 전환 이외에도 요청 값을 도메인 모델로 전환하는 등에도 사용되는 것을 봐왔는데 이름을 바꾸거나 다른 방식으로 모델을 전환해야할까요?

bactoria replied:
> domain 객체에서 다른 domain 객체로의 전환

이 부분은 응답용 모델이 아닌 도메인 모델이라고 착각했네요. 🥲

<br>

- `subway.db.h2.mapper` : JPA Entity <-> Domain Model
-  `subway.application.mapper`: Domain Model <-> Application Dto

로 생각하신거죠?

저는 mapper 라는 용어를 application layer 에서 사용하는 케이스를 보지 못해서 그런 것 같네요.
위의 mapper도 `subway.db.h2.mapper` 패키지에 위치한다고 생각했던 것 같아요.

Domain Model <-> Application Dto 변환을 어디선가 해줘야 하는데, 현재 Mapper 의 사용도 괜찮은 것 같습니다. 👍

toongri replied:
넵 알겠습니다 감사합니다!

bactoria commented on src/main/java/subway/domain/PathStation.java:
```diff
@@ -0,0 +1,60 @@
+
+@Getter
+@EqualsAndHashCode(of = {"id", "name"})
+public class PathStation {
```
경로 미션을 진행하시면서 `Station`를 활용하는 대신 `PathStation` 도메인 객체를 만들어주셨는데요!

`Station` 과 `PathStation`은 매우 유사해 보여요.
`PathStation` 으로 인해 복잡도가 증가한 것에 비해 느끼신 장점이 있으셨는지 궁금합니다ㅎㅎ

toongri replied:
아 저는 현재 기능이 적어서 Station과 PathStation이 유사해보일 수 있다고 생각합니다.
하지만 제 생각엔 현재 비슷해보이는 것이지 의미하는 바가 조금 다르다고 생각해 분리를 하게 되었는데요

지하철 역 도메인 모델(현재 Station)에 점점 다양한 정보가 들어가고 비즈니스 로직이 들어간다고 했을 때를 상상했을 때,
경로 조회 시에 역의 간략한 정보의 책임을 갖고 있는 경로 역 도메인 모델(현재 PathStation)에 그런 권한들이 존재할 필요가 없다 생각했습니다.

또한 경로 조회 시에도 많은 비즈니스 로직들이 포함되게 되면 지하철 역 도메인 모델은 점점 뚱뚱해지고, 모듈간 결합도가 높아져 분리가 어려워질 것이라고 판단했습니다.
언제든지 경로 모듈과 지하철 역 모듈이 쉽게 분리될 수 있도록 역이라는 비슷한 의미를 내포하고 있는 도메인 모델을 둘로 분리하게 되었습니다.

bactoria replied:
경로 모듈 / 지하철역 모듈 이렇게 두개의 모듈이라고 생각하신거군요.

경로 모듈 / 지하철역 모듈 로 생각하고 계시다면 
해당 클래스들을 같은 패키지에 두는 것 보다 domain 패키지 내부에 따로 패키지를 구분지어 두는 건 어떨까요?

모듈의 구분이 명확하여 의존성 관리하기가 더 좋을 것 같아요.

toongri replied:
아 맞습니다. 제가 그래서 패키지 분리를 해야겠다고 저도 판단을 했는데요.
그 판단이 너무 늦었습니다. 커밋을 몇개 쌓아놓고 의존성 관리가 된 채로 아 패키지를 분리해야겠다 마음을 먹었더니 꼬이는게 한두가지가 아니더라구요ㅠ

아마 다음 과제에서 분리된 채로 하지 않을까 생각하고 있습니다.

bactoria reviewed (changes requested):
준형님 안녕하세요!
바로 리뷰 드리려고 했는데 생각보다 늦어졌네요. 🙏

구현 잘 해주셨는데
몇가지 의견이 있어서 남겨두었어요. 한번 확인 부탁드릴게요!

다음 미션은 아직 진행 전이신 것 같은데, 
이번 코멘트는 부담 갖지 마시고 의견 주세요!ㅎㅎ 💪💪

toongri committed:
구간 추가 매니저 Spring 종속성 제거

SubwaySectionAddService/constructor 수정
SectionAddManager/annotation 제거
SectionDefaultAdder/annotation 제거
SectionMiddleAdder/annotation 제거
SubwaySectionAddServiceTest/test 수정

toongri committed:
구간 비활성화 매니저 Spring 종속성 제거

SubwaySectionCloseService/constructor 수정
SectionCloseManager/annotation 제거
SectionDefaultCloser/annotation 제거
SectionMiddleCloser/annotation 제거
SectionTailCloser/annotation 제거
SubwaySectionCloseServiceTest/test 수정

toongri committed:
경로 최단거리 탐색 기능 테스트 수정

PathShortestDistanceSearcherTest 리팩토링(경로구간 생성 함수 추가, setUp함수 추가)

toongri committed:
경로 탐색 결과 도메인 모델 이름 수정

PathSubway -> SubwayPath로 수정

bactoria reviewed (changes requested):
준형님 안녕하세요. 
제가 주말부터 화요일까지 여행을 다녀오느라 정신이 없어 리뷰도 늦어졌네요 😭

의견 남겨주신거에 대해 코멘트를 달아두었어요.
확인 후 다시 리뷰요청 주시면 머지하도록 할게요!! 💪💪💪

toongri committed:
경로 검색 인수 테스트 존재하지 않는 역 탐색 추가

PathSearchAcceptanceTest/searchPathBetweenNotExistStations 추가

bactoria commented on src/test/java/subway/acceptance/PathSearchAcceptanceTest.java:
```diff
@@ -108,4 +108,31 @@ public void searchPathBetweenDuplicatedSection() {
+     * @then 최단 거리로 갈 수 있다.
+     */
+    @Test
+    @DisplayName("존재하지 않은 역을 기준으로 경로 탐색을 할 수 없다.")
```
👍👍

bactoria reviewed (approved):
준형님 미션 진행하느라 고생 많으셨습니다. 👏👏👏
다음 미션들 진행하시면서 통찰력을 많이 얻어가시면 좋겠습니다.

화이팅입니다. 😄
